### DISCUSSION TITLE
Strange behaviour - is this by design?

### DISCUSSION THREAD

**Discussion by bezner** [pmndrs/jotai#1693] [Q&A]
Hey all,

First of all I need to mention that I have very small experience with react / react native so please forgive if my question is naive to you ;) 

I build a simple app which needs keep the user state in a storage once he's signed in through some API. To make it look simple I skip most of the irrelevant code just show you the most important parts. I have picked jotai state management package and I have a simple like this:

`<Provider>
            <Suspense fallback={<Text>Loading... </Text>}>
                <StatusBar style="dark"/>
                <Navigation/>
            </Suspense>
        </Provider>`
        
then I have the atom declared like so

`export const userAtom = atomWithStorage('userState', null, createJSONStorage(() => AsyncStorage));`

Which I use in Navigation component with: 

`const [user] = useAtom(userAtom);

    return (
        <NavigationContainer>
            <Stack.Navigator>
                { !user ? (
                    <>
                        <Stack.Screen name={'LogIn'} component={LogInScreen}/>
                        <Stack.Screen name={'Register'} component={RegisterScreen}/>
                        <Stack.Screen name={'PasswordReset'} component={PasswordResetScreen}/>
                    </>
                ) : (
                    <>
                        <Stack.Screen name="Tabs" component={TabsStack} options={{
                            headerShown: false
                        }}/>
                    </>
                ))}
            </Stack.Navigator>
        </NavigationContainer>
    );
`

where TabsStack is nested tabs navigation one of which pages is Profile screen where users can edit their profile. On that screen i have a log out button which resets the atom to its default value like so:

`onPress={() => setUser(RESET)}`

Now happens something unexpected. I would expect that the userAtom is reset to default value (null), the Navigation component conditionally renders authentication stack with LoginScreen. Instead, ProfileScreen component is re-rendered with user value set to null which causes many irresolvable errors like calling object properties of null (userAtom is an object). I really don't understand that and maybe that by design? Anyone please help me. 

I just mention that the AsyncStorage being used is indeed reset to null and when I restart the application once more - I have to login again. That being said I really don't know why after atom is reset to default value, the ProfileScreen is reloaded, not the whole stack.

**Comment by dai-shi** [comment]
Seems like you are using react-navigation or something. I wonder how Provider would work in this case. Can you try removing `<Provider>`? It will work with a default store (aka provider-less mode). If that doesn't change the behavior, my guess is wrong and need to revisit.

**Reply by bezner** [reply]
Thank you for your help @dai-shi but unfortunately the behaviour is exactly the same even though, like you said, the atoms work and the state persists (in my case). Yes I'm using react navigation v6 but the stack is really simple. So is the app.

**Reply by bezner** [reply]
One more thing, which makes me thinking I simply miss something, before I implemented your package, I tried recoil - the latest version - with the same result. Maybe it's react navigation issue than, I don't even know how could I test without it?

I'm new to react / react native but to my understanding with the state / atom defined like so

`export const userAtom = atomWithStorage('userState', null, createJSONStorage(() => AsyncStorage));`

each component using that state should react to state changes, with that said, my Navigation component shouldn't even have TabsStack in navigation stack, just the login screen etc. I'm lost.

**Reply by bezner** [reply]
I managed to find a temporary fix to this by creating a separate LogOutScreen component. Then user after pressing Log out button is navigated to that screen. The simple UseEffect makes the trick, like so:

```
const setUser = useSetAtom(userAtom);
useEffect(() => {
        setUser(RESET);
    }, []);
```
  
Shouldn't it work the other way as well though?

**Reply by dai-shi** [reply]
How do you set user on login in your code?

**Reply by bezner** [reply]
Hey @dai-shi, I submit input on button press

```
<Button color={'primary'} onPress={login}>
   Sign In
</Button>
```
than handle it like this
```
const login = () => {
        axios.post(environment.api_url + '/login', {
            email: email,
            password: password
        }).then((response) => {
            if (response && response.data.error) {
                // display error todo: consider removing
            } else if (response && !response.data.error) {
                setUser(response.data);
            }
        }).catch((error) => {
            console.error(error);
            // handle validation errors and such
        });
    }
```

That part seems to work. Anything on your mind?

**Reply by dai-shi** [reply]
Seems good.

Wait, you need to do `setUser(null)` instead of `useUser(RESET)`?

**Reply by bezner** [reply]
Hey, I tried both @dai-shi but none of those work. That's really strange. 

I'll try to isolate the code to a minimal example and send that to you later this week. Maybe there's some conflict with some other package? React navigation itself maybe? I don't know.

**Reply by dai-shi** [reply]
It's best if you can reproduce it with codesandbox + web. https://codesandbox.io/s/react-new
Otherwise, it's RN specific or react-navigation specific.

**Reply by bezner** [reply]
I'll try that.
